### PR TITLE
Add a bank conflict simulator

### PR DESCRIPTION
**Pull request by giuseros**
The idea is to use this tool to have a quick understanding of the bank conflicts that a given configuration will experience. It is also a good way to prototype LDS improvements withouth having to deal with the internals of the rocMLIR compiler.

### CODE REVIEW

**Comment by jerryyin** [comment]
Could you put a few case studies on 
 - how to invoke this script and 
 - how you figure out our original layout was bad on bank conflict using this script
 - how you use this script to figure out the layout to minimize bank conflict

I feel like the best place for this may be in the discussions: https://github.com/ROCmSoftwarePlatform/rocMLIR/discussions

**Comment by giuseros** [comment]
Hi @jerryyin , thanks for the push, I wanted to write about this for a while, but I was procrastinating :) 

Since LDS bank conflicts discussions tight a bit into the hardware, I would start by showing the simulator in a confluence page:
https://confluence.amd.com/display/MLSE/LDS+bank+conflict+simulator

If we agree that what I wrote can be public, I will upgrade to a discuss post.

**Comment by jerryyin** [comment]
@giuseros Thanks for the work. The confluence page look great for now. I hope this will be good study material to the team and reviewers!

File: mlir/utils/widgets/bank_conflicts.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+            for k in range(0, 1):
+                wavesToBanks[wave][m][k] = []
+
+    # View of the lds in terms of global offset
+    ldsToOffset = np.zeros([config.Kouter, config.D], dtype=np.int32)
+
+    if isKMajor:
+        for tid in range(0, config.blockSize):
+            for m in range(0, config.copyDPerThread):
+                waveid = tid // 64
```
Comment (giuseros) [inline mlir/utils/widgets/bank_conflicts.py]: Add a waveSize parameter

Comment (krzysz00) [reply mlir/utils/widgets/bank_conflicts.py]: Yep

File: mlir/utils/widgets/bank_conflicts.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+            config, parsed_args.kmajor, parsed_args.no_shuffle
+        )
+
+    if parsed_args.show_conflicts:
+        printBanks(config, wavesToOffset)
+
+    if parsed_args.show_offsets:
+        (_, ldsToOffset) = computeWriteBankConflicts(
+            config, parsed_args.kmajor, parsed_args.no_shuffle
+        )
```
Comment (krzysz00) [inline mlir/utils/widgets/bank_conflicts.py:318]: Is it just me or is that weird indentation

File: mlir/utils/widgets/bank_conflicts.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+                loffset = lrow * config.D + lcol
+                wavesToBanks[waveid][m][k].append(loffset * config.Kpack)
+    return wavesToBanks
+
+
+def main(args=None):
+    parser = argparse.ArgumentParser(
+        prog="Bank Conflicts simulator",
+        allow_abbrev=False,
+    )
```
Comment (krzysz00) [inline mlir/utils/widgets/bank_conflicts.py:243]: Same nit

**Review by krzysz00 — APPROVED** [review]
Approved, style nits

File: mlir/utils/widgets/bank_conflicts.py (hunk 2 above)
Comment (giuseros) [reply mlir/utils/widgets/bank_conflicts.py:318]: I used `black` to indent it, let me try it again :)

**Pull request merged**